### ISSUE TITLE
resize_image causing random build errors

### ISSUE THREAD

**Issue by jimmyff** [Keats/gutenberg#2079]
# Bug Report

Environment: Mac os 12.6.2

Zola version: 0.16.0

## Issue

Zola keeps getting itself in a state whereby it won't run. I get resize_image fail errors saying it can't find random images in my project. The only fix is to open the template that has the resize_image call, comment it out and then try again. Eventually the site will build again and I can go around and remove the comments and reintroduce the resize_image calls.

My site has a lot of images and a lot of resize_image calls. It takes around 20-30 seconds to build. (On a side note it would be great if some sort of resize_image caching could speed up the builds). 

```
jimmyff@jimmys-mbp-5 jimmyff % zola serve --drafts
Building site...
Error: Failed to serve the site
Error: Failed to render content of /Users/jimmyff/dev/jimmyff/content/gallery/2019-06-13-porcelain-wet-palette/index.md
Error: Reason: Failed to render image shortcode
Error: Reason: Failed to render 'shortcodes/image.html'
Error: Reason: Function call 'get_image_metadata' failed
Error: Reason: `get_image_metadata`: Cannot find path: .//gallery/porcelain-wet-palette/porcelain_wet_palette.webp
jimmyff@jimmys-mbp-5 jimmyff % zola serve --drafts
Building site...
Error: Failed to serve the site
Error: Failed to render content of /Users/jimmyff/dev/jimmyff/content/gallery/2022-05-10-still-life-whisky/index.md
Error: Reason: Failed to render image shortcode
Error: Reason: Failed to render 'shortcodes/image.html'
Error: Reason: Function call 'get_image_metadata' failed
Error: Reason: `get_image_metadata`: Cannot find path: .//gallery/still-life-whisky/whisky.webp
jimmyff@jimmys-mbp-5 jimmyff % zola serve --drafts
Building site...
Error: Failed to serve the site
Error: Failed to render content of /Users/jimmyff/dev/jimmyff/content/gallery/2019-06-19-sketch-a-day-painter/index.md
Error: Reason: Failed to render image shortcode
Error: Reason: Failed to render 'shortcodes/image.html'
Error: Reason: Function call 'get_image_metadata' failed
Error: Reason: `get_image_metadata`: Cannot find path: .//gallery/sketch-a-day-painter/jimmy.webp
jimmyff@jimmys-mbp-5 jimmyff % zola serve --drafts
Building site...
Error: Failed to serve the site
Error: Failed to render content of /Users/jimmyff/dev/jimmyff/content/gallery/2019-06-16-sketch-a-day-magical/index.md
Error: Reason: Failed to render image shortcode
Error: Reason: Failed to render 'shortcodes/image.html'
Error: Reason: Function call 'get_image_metadata' failed
Error: Reason: `get_image_metadata`: Cannot find path: .//gallery/sketch-a-day-magical/magical.webp
jimmyff@jimmys-mbp-5 jimmyff % 
```
Here you can see everytime i try to serve the site i get another random error. I'm not changing anything inbetween these attempts. The only fix seems to be to comment out the template and then add it back it once the build is working.

## Step to reproduce

I'm currently migrating my site from hugo to zola, I will be able to share the repo in the next day or two once I commit & push.

Thanks

**Comment by jimmyff** [comment]
So it seems to happen when a build was interrupted (`CMD+C`) when using `serve`, I think whatever `resize_image(..)` it was processing when interrupted is the template that breaks.

I only started running in to this issue as my site took longer and longer to build. I usually innterupt the build if I do soemthing like change a tag across 40 files, which then triggers 40 builds, each taking half a minute to complete.

**Comment by Keats** [comment]
> (On a side note it would be great if some sort of resize_image caching could speed up the builds).

There is? Processed images are saved in `/static/processed_images` with some hash in the filename and the images are not processed again if the output would match the hash. Eg in the docs there is https://github.com/getzola/zola/tree/master/docs/static/processed_images which contains all the resized images. Any fresh build on the docs site will not re-trigger the resizing.

As for the error, maybe the processing of the image is interrupted and the path returned doesn't contain the image yet?

**Comment by jimmyff** [comment]
> > (On a side note it would be great if some sort of resize_image caching could speed up the builds).

> There is? Processed images are saved in `/static/processed_images` with some hash in the filename and the images are not processed again if the output would match the hash. Eg in the docs there is https://github.com/getzola/zola/tree/master/docs/static/processed_images which contains all the resized images. Any fresh build on the docs site will not re-trigger the resizing.

I knew the resized images files were on the filesystem but as the build times still takes 30 seconds per change I assumed it was actually regenerating the files again in order to generate those hashes only to discover that they are already generated. 

Shouldn't my builds be taking 30 seconds then? 

> As for the error, maybe the processing of the image is interrupted and the path returned doesn't contain the image yet?

I'm not sure, I know that for the top error that mentions `/gallery/porcelain-wet-palette/porcelain_wet_palette.webp` I know this path exists in the `content/` folder but I'm not sure if it's lookin in the `public/` folder at this point of template generation. This is the error I'm reporting, I'm afraid I'm not sure about the fix. Would my repo be useful as a way of testing / reproducing it?

**Comment by c-git** [comment]
Yes if the repo is public it would help with reproducing the problem. I have been wondering why it takes 30 seconds wouldn't mind checking to see why.

**Comment by Keats** [comment]
Usually a slow build like that is caused by some Tera function calls in the parent template, which then happens for every page/section.

**Comment by jimmyff** [comment]
Hey, it's not ready to upload to a repo just yet as I'm still making some changes and sorting out my content so I've just zipped it up so you can hopefully see:
[jimmyff 2.zip](https://github.com/getzola/zola/files/10482317/jimmyff.2.zip)

Just adding reponsive images to the gallery which aren't working quite right yet.

Each build taking so long and not being able to interrupt it without getting the whole thing in an erronous state is really slowing things down. Any advice would be great thanks.

**Comment by Keats** [comment]
Thanks for the site. I've already found one unrelated issue in Zola while using it so that's good.
Your site is much slower than what I'm expecting based on the its size, I'll investigate more when I get some time.

**Comment by Keats** [comment]
With https://github.com/getzola/zola/pull/2086 you can now change your shortcodes to do:

```
{% set abspath = page.colocated_path ~ src %}
```

rather than guessing or doing some weird string manipulations.

The build for that site is down to 17s for me after the initial build that took around 27s. Still slow, I'll investigate a bit more.

**Comment by Keats** [comment]
Down to around 11s now using `page.colocated_path` after replacing it in the macros as well. Most of the time now is spent rendering taxonomies which is a different issue.

**Comment by jimmyff** [comment]
That's awesome thanks @Keats, for me to use the `colocated_path` path stuff would I need to build from source or is there an easier way? Currently installed using brew

**Comment by Keats** [comment]
Not possible without building from source for now.

**Comment by c-git** [comment]
Fortunately with cargo that is actually pretty easy. It just takes a while. You just need to clone the repo on the branch you need and then run `cargo build --release`. It took under 5 minutes on my computer.

**Comment by jimmyff** [comment]
I don't think this suggetion would would resolve the original issue of build errors if you interrupt the build during an image resize function. However a nice enhancement would be to allow you to continue browsing the last built website version whilst a new version is building? That would probably make the builds feel much shorter.

Sometimes a build happens almost instantly after a build finishes making it tricky to preview changes when there are long build times.

**Comment by Keats** [comment]
> However a nice enhancement would be to allow you to continue browsing the last built website version whilst a new version is building?

Most sites will rebuild on change in less than a second so I don't know if it's worth the effort.

> Sometimes a build happens almost instantly after a build finishes making it tricky to preview changes when there are long build times.

This is because we sometimes generate multiple change events for different content that could be batched. Hopefully it gets addressed during https://github.com/getzola/zola/issues/2077

**Comment by jkbecker** [comment]
I am getting failed `resize_image` errors on a tiny project that I just set up, complaining about an image that is definitely there.

This is the site repo: https://github.com/jkbecker/jkbecker.net

When I try building the site in Vercel with the default Zola settings (it just runs `zola build` on the repo root), I'm getting the following build fail:

````
[23:10:01.221] Running build in Washington, D.C., USA (East) – iad1
[23:10:01.278] Cloning github.com/jkbecker/jkbecker.net (Branch: main, Commit: fb7bf86)
[23:10:01.592] Previous build cache not available
[23:10:02.133] Cloning completed: 854.736ms
[23:10:02.305] Running "vercel build"
[23:10:02.882] Vercel CLI 28.14.1
[23:10:03.573] Building site...
[23:10:03.589] -> Creating 5 pages (0 orphan), 1 sections, and processing 0 images
[23:10:03.593] Failed to build the site
[23:10:03.593] Error: Failed to render page '/vercel/path0/content/contacts.md'
[23:10:03.593] Reason: Failed to render 'contact-page.html' (error happened in 'base.html').
[23:10:03.593] Reason: Function call 'resize_image' failed
[23:10:03.593] Reason: `resize_image`: Cannot find path: static/img/jkbckr.png
[23:10:03.608] Error: Command "zola build" exited with 1
````

As you can see from my repo, that file is definitely there (and it does reliably build locally for me)... it makes no sense to me?

**Comment by jimmyff** [comment]
@jkbecker my temporary workaround for this issue is: 

1. Delete the image_resize code from your template, macro or shortcode
2. Re-run the build/serve command
3. Once it's sucesfully rebuilt without the image reizing code then add the resize code back in
4. Run the build/serve command again and you should be fixed (at least until you trigger it again by inerrupting a build)

It's a pain but at least gets you back up and running. Hopefully the image resizing optimisations in the next version of zola will fix or at least make this much less likley to happen

**Comment by jkbecker** [comment]
Thanks for these instructions, but that won’t help my situation. It works just fine locally, but breaks in the automatic build on vercel.com where it’s not trivial or even possible to fiddle with the project like that…

**Comment by Keats** [comment]
It sounds like paths are different in Vercel then if it works locally?

**Comment by jimmyff** [comment]
Or perhaps your using git LFS and not setup the vercel build action to support it?

**Comment by jkbecker** [comment]
Not using LFS and both in the local case as well as Vercel I'm just calling `zola build` from the repo root.
The only difference I can see is that locally I'm running Zola 0.16.1 and Vercel is running 0.13.0... Would that explain my issue (i.e. another problem with `resize_image` that has been fixed between 0.13 and 0.16)?

**Comment by Keats** [comment]
Oh yeah the different versions would do it. The path handling has been rewritten in between those versions.

**Comment by jimmyff** [comment]
@Keats I've been using 0.17 and all seems good. Big improvement for the build time of my site. I no longer encounter the issue I originally discribed now that I'm using the new `colocated_path` -so happy for you to close this issue whenever you want. There is one little bug I've encoutered which is with `get_url` but I've rasied a seprate issue for that.

Build times for me now I've moved all my content over is 25-30 second (184 pages). I'll be uploading my site to github over the next couple of days, do you want me to open a new issue with it's repo link when I do, so you can use it for testing any future optimisations? (you mentioned something about my site not building as quickly as you expected, maybe taxonomy related?).

Thanks

**Comment by Keats** [comment]
> do you want me to open a new issue with it's repo link when I do

yes please

**Comment by jimmyff** [comment]
This is resolved in 0.17
